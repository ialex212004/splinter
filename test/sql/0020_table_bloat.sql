-- drop existing schema if it exists
drop schema if exists test_bloat cascade;

-- create schema and table
create schema test_bloat;
create table test_bloat.bloat_test (
    id serial primary key,
    data text,
    padding char(5000) -- ensure very wide rows to amplify bloat
);

-- disable autovacuum for consistency
alter table test_bloat.bloat_test set (autovacuum_enabled = false);

-- generate +90% bloat so this test doesn't fail randomly
insert into test_bloat.bloat_test (data, padding)
select 'Sample data ' || generate_series(1, 1000), repeat('x', 5);

-- delete 90% of the rows
delete from test_bloat.bloat_test where id % 10 != 0;

-- insert 100k new rows to ensure file growth
insert into test_bloat.bloat_test (data, padding)
select 'New data ' || generate_series(1, 1000), repeat('y', 3000);

-- delete 90% of the rows
delete from test_bloat.bloat_test where id % 10 != 1;

-- analyze the table to update statistics
analyze test_bloat.bloat_test;

-- expect 1 failure
select * from lint."0020_table_bloat";

-- resolve the issue
vacuum  test_bloat.bloat_test;
analyze test_bloat.bloat_test;

-- expect 0 failures
select * from lint."0020_table_bloat";

drop schema test_bloat cascade;



