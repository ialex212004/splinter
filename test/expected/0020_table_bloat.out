-- drop existing schema if it exists
drop schema if exists test_bloat cascade;
NOTICE:  schema "test_bloat" does not exist, skipping
-- create schema and table
create schema test_bloat;
create table test_bloat.bloat_test (
    id serial primary key,
    data text,
    padding char(5000) -- ensure very wide rows to amplify bloat
);
-- disable autovacuum for consistency
alter table test_bloat.bloat_test set (autovacuum_enabled = false);
-- generate +90% bloat so this test doesn't fail randomly
insert into test_bloat.bloat_test (data, padding)
select 'Sample data ' || generate_series(1, 1000000), repeat('x', 5);
-- delete 90% of the rows
delete from test_bloat.bloat_test where id % 10 != 0;
-- insert 100k new rows to ensure file growth
insert into test_bloat.bloat_test (data, padding)
select 'New data ' || generate_series(1, 1000000), repeat('y', 3000);
-- delete 90% of the rows
delete from test_bloat.bloat_test where id % 10 != 1;
-- analyze the table to update statistics
analyze test_bloat.bloat_test;
-- expect 1 failure
select * from lint."0020_table_bloat";
    name     |    title    | level |  facing  |  categories   |                                                 description                                                  |                       detail                        |                                                remediation                                                |                            metadata                             |             cache_key             
-------------+-------------+-------+----------+---------------+--------------------------------------------------------------------------------------------------------------+-----------------------------------------------------+-----------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------+-----------------------------------
 table_bloat | Table Bloat | INFO  | EXTERNAL | {PERFORMANCE} | Detects if a table has excess bloat and may benefit from maintenance operations like vacuum full or cluster. | Table `test_bloat`.`bloat_test` has excessive bloat | Consider running vacuum full (WARNING: incurs downtime) and tweaking autovacuum settings to reduce bloat. | {"name": "bloat_test", "type": "table", "schema": "test_bloat"} | table_bloat_test_bloat_bloat_test
(1 row)

-- resolve the issue
vacuum full test_bloat.bloat_test;
analyze test_bloat.bloat_test;
-- expect 0 failures
select * from lint."0020_table_bloat";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

drop schema test_bloat cascade;
NOTICE:  drop cascades to table test_bloat.bloat_test
